### PR TITLE
feat: disable SELFDESTRUCT opcode in l2geth

### PR DESCRIPTION
**Pull request by Thegaram**

### CODE REVIEW

**Comment by ChuhanJin** [comment]
Can one of the admins verify this patch?

**Comment by mask-pp** [comment]
> Can one of the admins verify this patch?

I'm glad to review it.

**Comment by mask-pp** [comment]
> > Can one of the admins verify this patch?
> 
> I'm glad to review it.

**Comment by mask-pp** [comment]
Please delete `SELFDESTRUCT` opcode in `vm/opcode.go` and in some test cases.

File: core/vm/eips.go [hunk 1]
```diff
@@ -142,10 +142,9 @@ func enable2929(jt *JumpTable) {
 	jt[DELEGATECALL].constantGas = params.WarmStorageReadCostEIP2929
 	jt[DELEGATECALL].dynamicGas = gasDelegateCallEIP2929
 
-	// This was previously part of the dynamic cost, but we're using it as a constantGas
-	// factor here
-	jt[SELFDESTRUCT].constantGas = params.SelfdestructGasEIP150
-	jt[SELFDESTRUCT].dynamicGas = gasSelfdestructEIP2929
+	// SELFDESTRUCT is disabled in Scroll
```
Comment (mask-pp) [inline core/vm/eips.go:145]: Why not delete it directly?

File: core/vm/eips.go [hunk 2]
```diff
@@ -154,7 +153,9 @@ func enable2929(jt *JumpTable) {
 // - Reduces max refunds to 20% gas
 func enable3529(jt *JumpTable) {
 	jt[SSTORE].dynamicGas = gasSStoreEIP3529
-	jt[SELFDESTRUCT].dynamicGas = gasSelfdestructEIP3529
+
+	// SELFDESTRUCT is disabled in Scroll
```
Comment (mask-pp) [inline core/vm/eips.go:157]: ditto

File: core/vm/jump_table.go [hunk 3]
```diff
@@ -1021,13 +1021,8 @@ func newFrontierInstructionSet() JumpTable {
...
 		},
-		SELFDESTRUCT: {
-			execute:    opSuicide,
-			dynamicGas: gasSelfdestruct,
-			minStack:   minStack(1, 0),
-			maxStack:   maxStack(1, 0),
-			halts:      true,
-			writes:     true,
-		},
+		// SELFDESTRUCT is disabled in Scroll.
```
Comment (mask-pp) [inline core/vm/jump_table.go:1024]: ditto

File: core/vm/eips.go (hunk 1 above)
Comment (icemelon) [reply core/vm/eips.go:145]: No, we should not delete it in case we may support it in the future and it'll incur more incompatiblity.

**Review by icemelon — COMMENTED** [review]
I think it'll be better if we fail the transaction during the contract deployment if the contract contains `SELFDESTRUCT` opcode.

**Comment by Thegaram** [comment]
> I think it'll be better if we fail the transaction during the contract deployment if the contract contains `SELFDESTRUCT` opcode.

I think that approach cannot prevent this opcode from being deployed in all cases. A contract could programmatically generate some bytecode that has `SELFDESTRUCT` and deploy it as a new contract. So if the VM encounters this opcode, we'll need to deal with it somehow, and I think treating it as an invalid instruction is the cleanest solution.

Also, in `tx.data`, how would we differentiate `SELFDESTRUCT` from other legitimate uses of `0xff`? E.g. when it's used as an `uint256` or as a byte in some arbitrary `bytes` data.

File: core/vm/eips.go (hunk 2 above)
Comment (Thegaram) [reply core/vm/eips.go:157]: The lines commented out are part of `EIP-2929` and `EIP-3529`. I think it's better to be explicit about disabling them here.

File: core/vm/jump_table.go (hunk 3 above)
Comment (Thegaram) [reply core/vm/jump_table.go:1024]: Same here, I'd prefer to disable this opcode more explicitly, if this only costs us this one line.

**Comment by 0xmountaintop** [comment]
close & reopen to trigger CI

File: core/vm/eips.go (hunk 1 above)
Comment (mask-pp) [reply core/vm/eips.go:145]: ok, that's great.

**Review by 0xmountaintop — APPROVED** [review]

**Review by icemelon — APPROVED** [review]
LGTM

**Pull request merged**